### ISSUE TITLE
Allow plugins for file connectors - Separation of concerns

### ISSUE THREAD

**Issue by ajbogh** [inadarei/connect-thumbs#5]
Currently the connect-thumbs.js file handles all URL decoding, file downloading and creation, and image processing. While it allows for a different callback method for decoding the URL it doesn't allow someone to modify the downloading and file creation methods.

This proposal will be to allow a developer to modify the various aspects with plugins. An example structure is below. Alternatively, one could use a node plugin instead of creating a file.
- /lib/connect-thumbs.js
  - plugins/base64-url-decode.js : decodes base64-encoded parameter into a URL
  - plugins/http-file-download.js : downloads file from external website
  - plugins/filesystem-retriever.js : retrieves file from the file system, instead of a URL

I would imagine that these plugins would need to be passed as parameters to override the various methods but I haven't thought about how to link that part together.

**Comment by inadarei** [comment]
Sounds interesting. Thanks

**Comment by ajbogh** [comment]
After a bit of work I've refactored some code into a new project based on connect-thumbs. 

I added some safety measures and the ability to pull original files from the file system (place images in the app/images directory or symlink app/images to another directory). With this update I've also had to change the API by creating a ?base64 query string parameter, width, w, height, h, and quality parameters.

I hope these changes can be improved on and incorporated into this project along with some of the suggestions I had above.

https://github.com/ajbogh/node-image-farmer
